ambry-router/src/main/java/com.github.ambry.router/PutOperation.java

```diff
@@ -1135,6 +1152,34 @@ private void finalizeMetadataChunk() {
     }
 
     /**
+     * if this is a composite object, fill the list with successfully put chunk ids and their {@link BlobProperties}
+     * @return the list of pairs of successfully put chunk ids and chunk and their {@link BlobProperties} if this is a
+     *         composite object, empty list otherwise.
+     */
+    List<Pair<StoreKey, BlobProperties>> getSuccessfullyPutChunkIdsAndPropertiesIfComposite() {
```

Comment:
This looks correct, but seems a bit more involved than I imagined. Also, I just realized that if we only initiate when operation completes, then we do not send notifications for successfully put chunk ids for a failed put. Let us discuss and resolve this today.